packages/vscode-extension/src/project/project.ts

```diff
@@ -838,19 +842,24 @@ export class Project
   }
 
   private reportStageProgress(stageProgress: number, stage: string) {
-    if (stage !== this.projectState.startupMessage) {
+    if (this.projectState.status !== "starting" || stage !== this.projectState.startupMessage) {
       return;
     }
     this.updateProjectState({ stageProgress });
   }
 
   private updateProjectState(newState: Partial<ProjectState>) {
+    // NOTE: this is unsafe, but I'm not sure there's a way to enforce the type of `newState` correctly
+    const mergedState: any = { ...this.projectState, ...newState };
```

Comment:
```suggestion
   const acceptableKeys = [...list of all posible keys]
    newState = Object.fromEntries(Object.entries(newState).filter((entry) => acceptableKeys.includes(entry[0])))
    const mergedState: any = { ...this.projectState, ...newState };
```
I think something like that would do the trik, wouldn't it?

Reply:
This only helps in that we skip copying entries from a "larger" type into a "smaller" one, but I don't think that's really an issue (since TS prevents us from accessing these properties anyway). The actual problem is that the typesystem doesn't correctly verify `newState` correctly updates the current state.
Currently, it's impossible to write this function's signature in a safe way, simply due to the fact that you want the `newState` to either:
- only include the fields that the resulting `ProjectState` might have
- include every field that the resulting `ProjectState` must have, if the `newState.status` is different from `this.projectState.status`
But the method type cannot know the type of `this.projectState`.
There are ways to implement this well, but they would require further changes in this file (everywhere that `this.updateProjectState` is called) that I didn't feel were worth the effort right now, since the behaviour is exactly the same as before.